### ISSUE TITLE
Use Shapeless in finch-core: Heterogenous Routers

### ISSUE THREAD

**Issue by vkostyukov** [finagle/finch#221]
See the fist comment.

**Comment by vkostyukov** [comment]
Just an idea of the top of my head. What if we can define a heterogeneous `Router[L <: HList]` like this:

```scala
val api: Router[String :: RequestReader[String] :: Service[Req, Rep] :: HNil] =
  Get / "a" /> "b" ::
  Get / "b" /> RequestReader.value("b") ::
  Get / "c" /> httpService
```

and then serve it with `Httpx` if there is an instance of `EncodeResponse[A]` available for every type `A` from the `HList` (polymorphic fold?):

```scala
Httpx.serve(":8081", ap)
```

**Comment by ImLiar** [comment]
I'll try my best and dig this way. Was looking for a chance to try shapeless in the real world

**Comment by travisbrown** [comment]
@ImLiar I'm aiming to have a WIP pull request up by tomorrow night. You're definitely welcome to start looking and we can compare notes then, though.

**Comment by ImLiar** [comment]
@travisbrown ok, its cool for me

**Comment by vkostyukov** [comment]
Great! Thank you @travisbrown, @ImLiar!

**Comment by vkostyukov** [comment]
I've created a separate ticket to track the progress on replacement the composite types with HList: #242.

**Comment by vkostyukov** [comment]
I can't stop thinking about heterogeneous routers. And I'm not quite sure I know how it should look lilke. But here are my current thoughts (according to #254).

I really like the colon `::` operator we have right now on `RequestReader` and I believe it should migrate to `Endpoint`s (a merged version of `Router` and `RequestReader`). So we should be able to write something like this:

```scala
val getUser: Endpoint[User] = (Get :: "users" :: path("id").as[Int] :: param("name")) { (id, name) =>
  User(id, name)
}
```

Note, that `::` is a universal compositor for endpoints (in the same meaning it works right now for `RequestReader`). So, we can't really use `::` for  _heterogenous routers_. In fact, I'm not quite sure we even need a _router_ here. How about just to have an implicit conversion from `HList` to `Service[HttRequest, HttpResponse]`, so we can do it like this:

````scala
val getUser: Endpoint[User] = ???
val getTicket: Endpoint[Ticket] = ???
val both: Endpoint[User] :: Endpoint[Ticket] :: HNil = getUser :: getTicket :: HNil

Httpx.serve(":8081", both)
// or
Httpx.serve(":8081", getUser :: getTicket :: HNil)
```

So I'm suggesting not to wrap `HList` with `Router` (or `Endpoint`) in terms of heterogeneous router ( since it doesn't really fit well into the nature of `HList`, which is a product type not sum type) but wrap an `Endpoint`s in `HList`, which sounds reasonable to me.

Does it make any sense?

**Comment by ImLiar** [comment]
Hi guys. Any progress with shapeless?

**Comment by vkostyukov** [comment]
Hi @ImLiar,

The current state is here: https://github.com/finagle/finch/pull/276. Hopefully, we will merge this in a couple of days.

**Comment by vkostyukov** [comment]
This is fixed by #276.
